### PR TITLE
Implement response paging to support autorest.android

### PR DESCRIPTION
**Pull request by JianpingChen**
Add generic page class to azure/android/core/util/paging;
Add synchronous and asynchronous page collection classes in azure/android/core/http/responsepaging;

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataRetriever.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Represents a synchronous paged data provider with fixed page size
+ * @param <T> items in the page
+ * @param <P> page of items
+ */
+public abstract class PagedDataRetriever<T, P extends Page<T>> {
+    private final int pageSize;
+
+    protected PagedDataRetriever(int pageSize) {
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataRetriever.java:17]: We can probably use annotations such as `@NonNull` for this in addition to checking for nullness.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Represents a collection of paged data with fixed page size where pages can be retrieved synchrounously
+ * @param <T> type of the items contained in the page
+ * @param <P> page that contains the items
+ */
+public class PagedDataCollection<T, P extends Page<T>> {
+    private final PagedDataRetriever<T, P> pagedDataRetriever;
+    private LinkedHashMap<String, P> pages = new LinkedHashMap<String, P>();
+    private String firstPageId;
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:18]: Not necessary for this release, but we could reduce the API surface by removing both `firstPageId` and `getFirstPageId()` and having users pass a null `pageId` to `getPage()` for getting the first page. Said first page can also be cached by using the `null` key.

Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:18]: This applies to all `PagedCollection`s and `PageRetriever`s in this PR.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+     * Get the page size
+     * @return size of the page
+     */
+    public int getPageSize() {
+        return pageSize;
+    }
+
+    /**
+     * Get first page in the collection
+     * @return a Map.Entry that contains both Response and the page
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java:37]: The Javadoc says this returns a Map.Entry but the signature indicates this is a `Response<P>`.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataRetriever.java (hunk 1 above)
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataRetriever.java:17]: This applies to all `PagedRetriever`s in this PR.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ * @param <T> type of the items contained in the page
+ * @param <P> page that contains the items
+ */
+public class PagedDataResponseCollection<T, P extends Page<T>> {
+    private final PagedDataResponseRetriever<T, P> pagedDataRetriever;
+    private LinkedHashMap<String, Response<P>> pages = new LinkedHashMap<String, Response<P>>();
+    private String firstPageId;
+
+    /**
+     *  Constructor requires a page provider that returns page data and response synchronously
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:25]: nit
```suggestion
     * Constructor requires a page provider that returns page data and response synchronously
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    /**
+     * Retrieves the first page in the collection synchronously
+     * @return page data along with the response for retrieving the first page
+     */
+    public Response<P> getFirstPage() {
+        if (firstPageId != null) {
+            return pages.get(firstPageId);
+        }
+        Response<P> firstPageResponse = pagedDataRetriever.getFirstPage();
+        firstPageId = firstPageResponse.getValue().getPageId();
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:41]: Is it guaranteed that `firstPageResponse` or `firstPageResponse.getValue()` won't be `null`?

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        pages.put(firstPageId, firstPageResponse);
+        return  firstPageResponse;
+    }
+
+    /**
+     * Retrueve a page synchronously with its id along with the response for retrieving it
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:47]: typo
```suggestion
     * Retrieve a page synchronously with its id along with the response for retrieving it
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+    /**
+     * Construction requires an asynchronous paged data provider
+     * @param pagedDataRetriever an asynchronous page data provider
+     */
+    public AsyncPagedDataCollection(AsyncPagedDataRetriever<T, P> pagedDataRetriever) {
+        this.pagedDataRetriever = pagedDataRetriever;
+    }
+
+    private void cacheResponse(P page, Response response ){
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:34]: nit
```suggestion
    private void cacheResponse(P page, Response response){
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        pages.put(page.getPageId(), new Map.Entry<Response, P>() {
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:35]: Do we expect that the value of a given page will change for a given `Response`? I feel that we should use a different data structure to represent this pair, we could even have our own custom immutable `SimplePair`.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/SimpleCallback.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
+package com.azure.android.core.http;
+
+/**
+ * Callback to receive a service operation result.
+ *
+ * @param <U> The type of the result.
+ */
+public interface SimpleCallback<U> {
```
[the remark this answers is not in the document]
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/SimpleCallback.java:8]: I'll rebase and use Callback instead. Thanks

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/SimpleCallback.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package com.azure.android.core.http;
```
[the remark this answers is not in the document]
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/SimpleCallback.java:1]: Done.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/Page.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    public Iterator<T> getIterator() {
+        return items.iterator();
+    }
+
+    /**
+     * Set next page id
+     * @param nextPageId id of the page next to this page
+     * @return current page
+     */
+    public Page<T> setNextPageId(String nextPageId) {
```
[the remark this answers is not in the document]
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/Page.java:58]: In the case of a pageable Http service, they normally don't set previous page id. For the ease of implementation of DataSource, we can internally set it in our caches

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/SimpleCallback.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/SimpleCallback.java]: deleted this class

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java (hunk 8 above)
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:35]: Good idea. I didn't like Map.Entry at all but Pair is only available for SDK version 24 (26?) or later and we'd like to support 21

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java (hunk 5 above)
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:41]: Good point. will check in a fix

File: eng/common/scripts/update-docs-metadata.ps1 [hunk 12]
```diff
@@ -70,12 +70,14 @@ function GetAdjustedReadmeContent($pkgInfo, $lang){
     }
 
     $fileContent = $pkgInfo.ReadmeContent
-    $foundTitle = ""
```
Comment (weshaggard) [inline eng/common/scripts/update-docs-metadata.ps1:73]: Looks like this is unrelated and might be a bad merge.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/Page.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/Page.java:62]: Sounds good, I wrote this comment before I saw how you use these in the different `PagedCollection`s.

File: eng/common/scripts/update-docs-metadata.ps1 (hunk 12 above)
Comment (JianpingChen) [reply eng/common/scripts/update-docs-metadata.ps1:73]: Yeah. I have no intension of changing this file.

**Comment by JianpingChen** [comment]
Sorry guys, looks like this branch is messed up. Let me create a new PR.
